hare/broker.go

```diff
@@ -154,22 +118,40 @@ func (b *Broker) handleMessage(ctx context.Context, msg []byte) error {
 	logger.Debug("broker received hare message")
 
 	msgLayer := hareMsg.Layer
-	if !b.Synced(ctx, msgLayer) {
-		return errNotSynced
+
+	if !(b.nodeSyncState.IsSynced(ctx) && b.nodeSyncState.IsBeaconSynced(msgLayer.GetEpoch())) {
+		err := fmt.Errorf("broker rejecting message because not synced")
+		logger.With().Debug("rejecting message", log.Err(err))
+		return err
 	}
+	b.mu.Lock()
+	out, exist := b.outbox[msgLayer.Uint32()]
+	if !exist {
+		logger.With().Debug(
+			"broker received a message to a consensus process that is not registered",
+			msgLayer.Field(),
+			log.Stringer("msg_type", hareMsg.InnerMsg.Type),
+		)
 
-	isEarly := false
-	if err = b.validateTiming(ctx, &hareMsg); err != nil {
-		if !errors.Is(err, errEarlyMsg) {
-			// not early, validation failed
-			logger.With().Debug("broker received a message to a consensus process that is not registered",
-				log.Err(err))
-			return err
+		// We consider messages for the next layer to be early and we process
+		// them, all other messages we ignore. We don't impose the outbox
+		// capacity at this point since this is an externally initiated event
+		// and at most it can cause us to exceed our capacity by one slot.
+		if msgLayer == b.latestLayer.Add(1) {
+			out = make(chan any, inboxCapacity)
+			b.outbox[msgLayer.Uint32()] = out
+		} else {
+			b.mu.Unlock()
```

Comment:
ditto. this is error-prone. please find a way such that locking/unlocking is done safely.
maybe move this block of code out to a separate method so you can access outbox and latestLayer under the same lock.

Reply:
Also wrapped in a function call